config.txt
MAP_CYCLE = map1.stl, map2.stl, map4.stl, map5.stl, map6.stl
LAPS = 1
ROUND_TIME = 10
MAX_PLAYERS = 16
SKIP_ROUNDS_CAR_CAR = 1
SKIP_ROUNDS_COLLISION_PER_10_KMH = 0.5
SKIP_ROUNDS_COLLISION_MIN = 1
PORT = 3410
COUNTDOWN = 60
WELCOME_MSG = Welcome!
SERVER_NAME = My Server


SERVER_SETTINGS.md
Setting up a GridCars server
============================================

Normal Server:
-----------------------
A normal server is one with a graphical interface (where you also play along). Usually used in LAN settings.
- Edit your config.txt file in _%APPDATA%\GridCars\_ or _~/.local/share/GridCars/_ (The exact path to the file is shown when you start the game and press 'H' for help.) Available options: see below
- If you want to play on the internet, port-forward the Port (Default: TCP 3140) on your router
- Start the Game, and select one of the two server options.

Dedicated server:
-----------------------
A dedicated server is run from the console, usually on online servers.
- Download the source code version of the game [from github](https://github.com/Germanunkol/GridCars/)
- Edit the config.txt file _in the root directory of the game_. Available options: see below
- Make sure Lua and luasocket are installed.
- Run "lua dedicated.lua" in the root directory.

Options:
----------------------
The server has the following options which you can set in the config.txt file:

- MAP_CYCLE: Comma-sperated list of the maps. These files must be in the maps/ subfolder. Names are case-sensitive.
- LAPS: Number of laps a player needs to finish in order to win.
- ROUND_TIME: Number of seconds a player has 
- MAX_PLAYERS: Number of players allowed to join this server. All standard maps support at least 16 players, maps 5 and 6 support at least 64 players.
- SKIP_ROUNDS_CAR_CAR: Number of rounds a player has to wait after crashing into another car.
- SKIP_ROUNDS_COLLISION_PER_10_KMH: After a crash, the greater the player's speed was, the longer they have to wait. This value tells the game how much to add to the waiting time, per 10 km/h of the player's speed. Set this to zero if you don't want the number of skipped rounds to depend on the player's speed.
- SKIP_ROUNDS_COLLISION_MIN: Minimum number of rounds player has to wait after crashing into the wall.
- PORT: TCP port to use. This port must be opened on your router if you want to play over the internet!
- COUNTDOWN: Number of seconds to wait before starting a new round, after at least one player is ready.
- WELCOME_MSG: Chat message sent to new players
- SERVER_NAME: Name of the server to display in serverlist.


Here's an example along with the default values:
```
MAP_CYCLE = map1.stl, map2.stl, map4.stl, map5.stl, map6.stl
LAPS = 0
ROUND_TIME = 10
MAX_PLAYERS = 16
SKIP_ROUNDS_CAR_CAR = 1
SKIP_ROUNDS_COLLISION_PER_10_KMH = 0.5
SKIP_ROUNDS_COLLISION_MIN = 1
PORT = 3410
COUNTDOWN = 60
WELCOME_MSG = "Welcome!"
SERVER_NAME = My Server
```


config.lua
-- Config file handling, can save and load config data
-- By Germanunkol

-- Uses the format:
-- name1 = value1
-- name2 = value2
-- ...
-- Avoid using spaces and "=" signs in the names.
-- name must be a string or number, value must be string or number. Example:
--[[
level = l11.dat
lastLevel = n46.dat
fullscreen = false

Usage:
config = require("config.lua")

config.setValue( "lie", "cake", "configFile.txt" )
print(config.getValue( "lie", "configFile.txt" ))	-- should print "cake".

]]--

local config = {}

local CONFIG_FILE = "config.txt"	-- default file name if none is given.

-- Saves a name, value pair in the file "filename".
-- If an entry with the same name exists, it's overwritten.
function config.setValue( name, value, filename )

	if DEDICATED then return end

	filename = filename or CONFIG_FILE	-- default to configfile
	
	if not name or not value == nil then
		print(name, value)
		error("Error: configFile.setValue got nil value or nil name.")
	end
	
	if not love.filesystem.isFile( filename ) then
		local file = love.filesystem.newFile( filename )
		file:open('w')	--create the file
		file:close()
	end
	
	if type(value) ~= "string" then
		value = tostring(value)
	end
	
	local file = love.filesystem.newFile( filename )
	local data
	if file then
		file:open('r')
		data = file:read()
		file:close()
	end
	
	if not data then
		data = ""
	end
	
	local newData = ""
	local found = false
	--print("full:")
	for line in data:gmatch("[^\r\n]+") do
		--print(line)
		s, e = string.find(line, name .. " = [^\r\n]*")
		if s then
			--data = string.gsub(data, name .. " = [^\r\n]+\r\n", name .. " = " .. value .. "\r\n")
			newData = newData .. name .. " = " .. tostring(value) .. "\r\n"
			found = true
		else
			newData = newData .. line .. "\r\n"
			--data = data .. name .. " = " .. value .. "\r\n"
		end
	end
	if not found then
		newData = newData .. name .. " = " .. tostring(value) .. "\r\n"
	end
	
	file = love.filesystem.newFile( filename )
	if file then
		file:open('w')
		file:write(newData)
		file:close()
		return true
	end
end

-- Load and return the value corresponding to the name from filename.
-- If file can't be found or name is 
function config.getValue( name, filename )
	
	filename = filename or CONFIG_FILE	-- default to configfile

	if not DEDICATED and not love.filesystem.isFile(filename) then
		--print("Could not find config file.", name, filename)
		return nil
	end

	local data
	if not DEDICATED then
		local ok, file = pcall(love.filesystem.newFile, filename )
		if ok and file then
			file:open('r')
			data = file:read()
			file:close()
		end
	else
		local f = io.open(filename, "r")
		if f then
			data = f:read("*all")
			f:close()
		end
	end
	if data then
		for k, v in string.gmatch(data, "([^ \r\n]+) = ([^\r\n]*)") do
			if k == name then
				if v == "false" then return false end
				if v == "true" then return true end
				return v
			end
		end
	end
	print("Value for '" .. name .. "' not found in config file. Using default.")
	return nil
end

function config.load()
	PLAYERNAME = config.getValue( "PLAYERNAME" ) or "Unknown"
	ROUND_TIME = tonumber(config.getValue( "ROUND_TIME" )) or 10
	LAPS = tonumber(config.getValue( "LAPS" )) or 1
	MAX_PLAYERS = tonumber(config.getValue( "MAX_PLAYERS" )) or 16
	TRAIL_LENGTH = tonumber(config.getValue( "TRAIL_LENGTH" )) or 500
	SKIP_ROUNDS_CAR_CAR = tonumber(config.getValue( "SKIP_ROUNDS_CAR_CAR" )) or 1
	SKIP_ROUNDS_COLLISION_PER_10_KMH = tonumber(config.getValue( "SKIP_ROUNDS_COLLISION_PER_10_KMH" )) or 0.5
	SKIP_ROUNDS_COLLISION_MIN = tonumber(config.getValue( "SKIP_ROUNDS_COLLISION_MIN" )) or 1
	PORT = tonumber(config.getValue( "PORT" )) or 3410
	ADDRESS = config.getValue( "ADDRESS" ) or "germanunkol.de"
	COUNTDOWN = tonumber(config.getValue( "COUNTDOWN" )) or 60
	WELCOME_MSG = config.getValue( "WELCOME_MSG" ) or "Welcome!"
	SERVER_NAME = config.getValue( "SERVER_NAME" ) or "Default Server"

	PLAYERNAME = PLAYERNAME:sub( 1, 28 )	-- max of 28 chars
	SERVER_NAME = SERVER_NAME:sub( 1, 50 )	-- max of 50 chars


	SKIP_ROUNDS_CAR_CAR = math.max( 0, SKIP_ROUNDS_CAR_CAR )
	SKIP_ROUNDS_COLLISION_PER_10_KMH = math.max( 0, SKIP_ROUNDS_COLLISION_PER_10_KMH )
	SKIP_ROUNDS_COLLISION_MIN = math.floor(math.max( 0, SKIP_ROUNDS_COLLISION_MIN ))

	if not DEDICATED then
		WIDTH = tonumber(config.getValue( "WIDTH" )) or love.graphics.getWidth()
		HEIGHT = tonumber(config.getValue( "HEIGHT" )) or love.graphics.getHeight()
		FULLSCREEN = config.getValue( "FULLSCREEN" ) or false
		local w, h, flags = love.window.getMode()
		if WIDTH ~= w or HEIGHT ~= h or FULLSCREEN ~= flags.fullscreen then
			assert(love.window.setMode( WIDTH, HEIGHT, {fullscreen = FULLSCREEN} ), "Cannot change window size. Change or remove your config.txt")
		end
	end

	-- Remove any pipe symbols from the player name:
	PLAYERNAME = string.gsub( PLAYERNAME, "|", "" )

	if not DEDICATED then
		config.createIfEmpty()
	end
end


-- If no config file was found, write it:
function config.createIfEmpty()

	-- Don't let dedicated server write:
	if DEDICATED then return end

	local contents = love.filesystem.read( "config.txt" )

	if not contents then
		config.setValue( "PLAYERNAME", PLAYERNAME )
		config.setValue( "ROUND_TIME", ROUND_TIME )
		config.setValue( "WIDTH", WIDTH )
		config.setValue( "HEIGHT", HEIGHT )
		config.setValue( "LAPS", LAPS )
		config.setValue( "MAX_PLAYERS", MAX_PLAYERS )
		config.setValue( "TRAIL_LENGTH", TRAIL_LENGTH )
		config.setValue( "PORT", PORT )
		config.setValue( "COUNTDOWN", COUNTDOWN )
		config.setValue( "FULLSCREEN", false )
		config.setValue( "WELCOME_MSG", WELCOME_MSG )
		config.setValue( "SERVER_NAME", SERVER_NAME )
		config.setValue( "SKIP_ROUNDS_CAR_CAR", SKIP_ROUNDS_CAR_CAR )
		config.setValue( "SKIP_ROUNDS_COLLISION_PER_10_KMH", SKIP_ROUNDS_COLLISION_PER_10_KMH )
		config.setValue( "SKIP_ROUNDS_COLLISION_MIN", SKIP_ROUNDS_COLLISION_MIN )
	end
end

return config
